sql: minor cleanup around portal pausability

yuzefovich opened the pull request:
This PR contains several commits that cleanup and harden code around pausable portals. See each commit for details.

Informs: https://github.com/cockroachlabs/support/issues/3463.
Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/155670)
<!-- Reviewable:end -->

yuzefovich committed:
sql: cleanup the code around pausable portals a bit

Extract a common helper function as well as remove redundant argument
for another function.

Release note: None

yuzefovich committed:
sql: clarify cleanup queue for pausable portals

We have several cleanup "stacks" for pausable portals model, but in
reality we execute the functions in the "forward" direction, matching
the behavior of a "queue", so this commit renames the struct
accordingly. Only `execStmtInOpenState` and `dispatchToExecutionEngine`
stages have multiple cleanup functions.

Additionally, it fixes the inverted order of executing two cleanup
functions in `dispatchToExecutionEngine` stage. Namely, previously we
performed `planTop.Close` _before_ `recordStatementSummary` which
doesn't match the behavior on the main path, outside the pausable
portals. We now defer appending the former so that it's executed after
the latter.

Deferring `planTop.Close` was necessary for the following commit which
is what prompted this one.

Release note: None

yuzefovich committed:
sql: fix minor bug around disabling portal pausability

I've been reviewing the pausable portal code, and I think I found
a minor bug that doesn't actually appear to have any consequences. What
happens is that if we created a pausable portal, but then it's executed
without a limit (i.e. `MaxRows` is not set on `Execute`), then we
disable pausability since the portal will run to completion. We do that
by modifying a _copy_ of `PreparedPortal` struct that is used throughout
the execution, yet we don't update the map that stores all portals. Note
that other nested things (like `portalPauseInfo`) are stored by
reference, so modifications to them are reflected in the copy stored in
the map, so I think things just happen to work out. Still, it seems
prudent to update the copy that we store in the map with the fact that
pausability was disabled.

A similar issue is present when we revoke the portal pausability - we
modify the copy of PreparedPortal and don't update the map. Additionally,
here we explicitly reset `planner.pausablePortal` to `nil` when revoking
pausability. I audited all cleanup functions, and they should behave
correctly: namely, we'll have nil'ed out `pauseInfo`, so the portal won't
be considered pausable, and the cleanup will be performed right away,
avoiding the pausauble portal cleanup stacks.

Release note: None

ZhouXing19 reviewed (approved):
Great catches, <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>! Thanks for the fix. I only have 2 non-blocking comments.

> @ZhouXing19 reviewed 2 of 2 files at r1, 2 of 2 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155670#-:-Oc14zBvBDt4V8nSg7bt:b-asrpwo)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @michae2)

___
*[`pkg/sql/prepared_stmt.go` line 234 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/155670#-Oc0uIO0AVD6XgnS_Ue-:-Oc0uIO0AVD6XgnS_Ue0:b-h5334q) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cbacb36d221cd2c7c68069bc07260c4f00febe7/pkg/sql/prepared_stmt.go#L234)):*
> ```Go
> 	// which means running the cleanup functions of the four structs in the
> 	// following order:
> 	//   - exhaustPortal.cleanup
> ```

Thanks for calling out the ordering of the clean up functions! This also rings the bell that this comment is not correct -- we actually do the cleanup for each layer in the reverse ordering, so here it should be: 
```
//  - resumableFlow.cleanup
//  - dispatchToExecutionEngine.cleanup
//  - execStmtInOpenState.cleanup
//  - execStmtInOpenState.cleanup
```
The implementation already follows this (see `portalPauseInfo.cleanupAll()` and other callsites of `resumableFlow.cleanup`), so this is just a comment fix. Ditto for the comment below for the error handling case as well. Sorry for my previous confusion!
___
*[`pkg/sql/conn_executor_exec.go` line 2767 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/155670#-Oc1-Vev45xR6N7ckh_j:-Oc1-Vev45xR6N7ckh_k:b-b4vjrx) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cbacb36d221cd2c7c68069bc07260c4f00febe7/pkg/sql/conn_executor_exec.go#L2767)):*
> ```Go
> 				defer func() {
> 					// We need to check whether pausable portals model is still
> 					// used since it might have been revoked below.
> ```

Good catch! 
Maybe I'm missing some historic details here, but it seems that we can consolidate the check for revocability below there within this closure as well, so that we don't need to have the redundant `defer planner.curPlan.close()`. As in, in the below part, we check 

```
		if len(planner.curPlan.subqueryPlans) == 0 &&
			len(planner.curPlan.cascades) == 0 &&
			len(planner.curPlan.checkPlans) == 0 &&
			len(planner.curPlan.triggers) == 0 {
```

But these information of `planner.curPlan` should have been populated with `ex.makeExecPlan(ctx, planner)`. 
Not to block this PR though, and happy to put up another PR for these following cleanup.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
sql: unify revoking portal pausability

Previously, we were revoking portal pausability in two spots: once for
not-read-only stmts and separately if we find any sub- or post-queries.
We can actually check both things right after having performed the
logical planning, which unifies the code a bit.

Release note: None

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Great audit, thank you for doing this!

> @michae2 reviewed 2 of 2 files at r1, 2 of 2 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155670#-:-Oc1wryQ1wl-niH1bZy3:bxrwgf3)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155670#-:-Oc26UEk-h7OJNfx5L-9:b-tlx7ko)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @michae2 and @ZhouXing19)

___
*[`pkg/sql/conn_executor_exec.go` line 2767 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/155670#-Oc1-Vev45xR6N7ckh_j:-Oc1xGZz55Ojij7BOUK2:b-jtgw89) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cbacb36d221cd2c7c68069bc07260c4f00febe7/pkg/sql/conn_executor_exec.go#L2767)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

Good catch! 
Maybe I'm missing some historic details here, but it seems that we can consolidate the check for revocability below there within this closure as well, so that we don't need to have the redundant `defer planner.curPlan.close()`. As in, in the below part, we check 

```
		if len(planner.curPlan.subqueryPlans) == 0 &&
			len(planner.curPlan.cascades) == 0 &&
			len(planner.curPlan.checkPlans) == 0 &&
			len(planner.curPlan.triggers) == 0 {
```

But these information of `planner.curPlan` should have been populated with `ex.makeExecPlan(ctx, planner)`. 
Not to block this PR though, and happy to put up another PR for these following cleanup.
</blockquote></details>

Nice, good point, I like this simplification, so I extracted it into another (now second) commit.
___
*[`pkg/sql/prepared_stmt.go` line 234 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/155670#-Oc0uIO0AVD6XgnS_Ue-:-Oc1xDYS1lAAbd8FeTPJ:b-5mc357) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cbacb36d221cd2c7c68069bc07260c4f00febe7/pkg/sql/prepared_stmt.go#L234)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

Thanks for calling out the ordering of the clean up functions! This also rings the bell that this comment is not correct -- we actually do the cleanup for each layer in the reverse ordering, so here it should be: 
```
//  - resumableFlow.cleanup
//  - dispatchToExecutionEngine.cleanup
//  - execStmtInOpenState.cleanup
//  - execStmtInOpenState.cleanup
```
The implementation already follows this (see `portalPauseInfo.cleanupAll()` and other callsites of `resumableFlow.cleanup`), so this is just a comment fix. Ditto for the comment below for the error handling case as well. Sorry for my previous confusion!
</blockquote></details>

Good catch, updated the comments.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
bors r+
